### PR TITLE
checking if useoidc env var is set

### LINKED ISSUE #329: Use OIDC as Env Variable
## Describe the bug

To be able to remove hard-coded use_oidc = {true, false} config, provider code should check if an environment variable is exists, like POWER_PLATFORM_USE_OIDC.

### To Reproduce

Before;

`provider "powerplatform" {
  use_oidc = true
}`

After (with POWER_PLATFORM_USE_OIDC environment variable is set);

`provider "powerplatform" {
}
`

**REMINDER: REMOVE SENSITIVE DATA SUCH AS SECRETS, USER NAMES, EMAILS, TENANT INFORMATION, ETC.**

```hcl
provider "powerplatform" {
}
```

## System Information

 - Provider Version: 2.3.1-preview
 - OS & Version: [e.g. Windows, Linux, MacOS]
 

## Contribution
Do you plan to raise a PR to address this issue?  YES / NO?

### PR DESCRIPTION
**Pull request by polatengin**
To be able to remove hard-coded `use_oidc = {true, false}` config from provider section, provider should be able to determine if a specific environment variable is set.

Before;

```terraform
provider "powerplatform" {
  use_oidc = true
}
```

After _(with `POWER_PLATFORM_USE_OIDC` environment variable is set)_;

```terraform
provider "powerplatform" {
}
```

closes #329

### CODE REVIEW

**Review by eduardodfmex — DISMISSED** [review]
Thank you for this change that will help to carry the OIDC.

**Review by ianjensenisme — DISMISSED** [review]
Why does the environment config trump the provider config? This is more of a "food for thought" question, as I've also added variable handling that uses this priority order, so not worth changing, but I wonder if we've gotten it backwards throughout the provider. I'm probably just not thinking of an obvious reason to do things in that priority order. Anywho, LGTM.

**Comment by polatengin** [comment]
> Why does the environment config trump the provider config? This is more of a "food for thought" question, as I've also added variable handling that uses this priority order, so not worth changing, but I wonder if we've gotten it backwards throughout the provider. I'm probably just not thinking of an obvious reason to do things in that priority order. Anywho, LGTM.

it's more like, the options that you can use to configure the plan/apply in the way you like.

so, instead of having how you authenticate hard-coded, you might want to configure it in the pipeline, based on the pipeline parameters, etc.

_quick note; this change doesn't give priority to environment config, we still check if the config exists in the provider section, if yes, we use it, if no, we check the environment variables._

File: internal/powerplatform/provider.go [hunk 1]
```diff
@@ -181,6 +181,14 @@ func (p *PowerPlatformProvider) Configure(ctx context.Context, req provider.Conf
 		clientSecret = config.ClientSecret.ValueString()
 	}
 
+	useOidc := false
+	_, envUseOidcExists := os.LookupEnv("POWER_PLATFORM_USE_OIDC")
+	if config.UseOidc.IsNull() {
+		useOidc = envUseOidcExists
+	} else {
+		useOidc = config.UseOidc.ValueBool()
+	}
```
Comment (mattdot) [inline internal/powerplatform/provider.go:190]: Why is "exists" in the name of this variable?  It's not a actually if it exists.  It's the value of the environment variable.
```suggestion
	_, envUseOidc := os.LookupEnv("POWER_PLATFORM_USE_OIDC")
	if config.UseOidc.IsNull() {
		useOidc = envUseOidc
	} else {
		useOidc = config.UseOidc.ValueBool()
	}
```

**Review by mattdot — CHANGES_REQUESTED** [review]

File: internal/powerplatform/provider.go (hunk 1 above)
Comment (polatengin) [reply internal/powerplatform/provider.go]: fair 👍 made the changes..

**Comment by ianjensenisme** [comment]
> > Why does the environment config trump the provider config? This is more of a "food for thought" question, as I've also added variable handling that uses this priority order, so not worth changing, but I wonder if we've gotten it backwards throughout the provider. I'm probably just not thinking of an obvious reason to do things in that priority order. Anywho, LGTM.
> 
> it's more like, the options that you can use to configure the plan/apply in the way you like.
> 
> so, instead of having how you authenticate hard-coded, you might want to configure it in the pipeline, based on the pipeline parameters, etc.
> 
> _quick note; this change doesn't give priority to environment config, we still check if the config exists in the provider section, if yes, we use it, if no, we check the environment variables._

Ah, yep, I stared at it too long and lost the plot: we check if the config value is blank and only then would we use the environment variable, that makes more sense to me. I do basically the same thing with envOidcRequestUrl and envOidcRequestToken . Thanks for following up.

**Review by mawasile — APPROVED** [review]

**Review by eduardodfmex — DISMISSED** [review]

**Review by mattdot — DISMISSED** [review]

**Review by ianjensenisme — APPROVED** [review]

**Comment by eduardodfmex** [comment]
LGTM.

**Review by eduardodfmex — APPROVED** [review]
LGTM.

**Pull request merged**